samples/kafka-streams-interactive-query/src/main/java/com/example/kafka/streams/music/KafkaStreamsInteractiveQueryApplication.java

```diff
@@ -0,0 +1,183 @@
+
+@SpringBootApplication(proxyBeanMethods = false)
+class KafkaStreamsInteractiveQueryApplication {
+
+	private static final Logger LOG = LoggerFactory.getLogger(KafkaStreamsInteractiveQueryApplication.class);
+
+	public static void main(String[] args) {
+		SpringApplication.run(KafkaStreamsInteractiveQueryApplication.class, args);
+	}
+
+	@EnableConfigurationProperties(ProductTrackerProperties.class)
+	static class InteractiveProductCountApplication {
+
+		static final String PRODUCT_TOPIC = "product-events";
+		static final String PRODUCT_COUNT_STORE_NAME = "product-count-store";
+		static final String TOP_TWO_PRODUCTS_STORE_NAME = "top-two-products-store";
+		static final String TOP_TWO_PRODUCTS_KEY = "top-two";
+
+		@Value("${spring.cloud.stream.kafka.streams.binder.configuration.state.dir:UNKNOWN}")
+		private String stateStoreDir;
+
+		private final ProductTrackerProperties productTrackerProperties;
+
+		InteractiveProductCountApplication(ProductTrackerProperties productTrackerProperties) {
+			this.productTrackerProperties = productTrackerProperties;
+		}
+
+		@PostConstruct
+		void showStateStoreInfo() {
+			LOG.info("Using KafkaStreams state.dir = {}", this.stateStoreDir);
+		}
+
+		@Bean
+		public Consumer<KStream<Object, Product>> process() {
+			return (input) -> {
+				// Accept product events that are configured to be tracked
+				final KStream<Integer, Product> productsByProductId = input
+						.filter((key, product) -> productTrackerProperties.getProductIds().contains(product.id()))
+						.map((key, value) -> new KeyValue<>(value.id, value));
+
+				Serde<Product> productSerde = new JsonSerde<>(Product.class);
+				Serde<ProductCount> productCountSerde = new JsonSerde<>(ProductCount.class);
+				Serde<TopTwoProducts> topTwoSerde = new JsonSerde<>(TopTwoProducts.class).noTypeInfo();
+
+				// Create a state store to track product counts
+				final KTable<Product, Long> productCounts = productsByProductId
+						.groupBy((productId, product) -> product, Grouped.with(productSerde, productSerde))
+						.count(Materialized.<Product, Long, KeyValueStore<Bytes, byte[]>>as(PRODUCT_COUNT_STORE_NAME)
+								.withKeySerde(productSerde)
+								.withValueSerde(Serdes.Long()));
+
+				// Compute the top two products and store updated result in 'top-two-products-store'
+				productCounts.groupBy((product, count) ->
+								KeyValue.pair(TOP_TWO_PRODUCTS_KEY, new ProductCount(product.id(), count)),
+								Grouped.with(Serdes.String(), productCountSerde))
+						.aggregate(
+								TopTwoProducts::new,
+								(aggKey, value, aggregate) -> aggregate.add(value),
+								(aggKey, value, aggregate) -> aggregate.remove(value),
+								Materialized.<String, TopTwoProducts, KeyValueStore<Bytes, byte[]>>as(TOP_TWO_PRODUCTS_STORE_NAME)
+										.withKeySerde(Serdes.String())
+										.withValueSerde(topTwoSerde)); //new TopTwoProductsSerde()));
+			};
+		}
+	}
+
+	record Product(Integer id) { }
```

Comment:
ℹ️ I ❤️ the Java14+ `record` which allows us to delete all of the boilerplate init/get/set. Jackson knows how to handle this out of the box in 2.13 👍🏻